Review of simpoints work

BobbyRBruce opened the pull request.

studyztp committed:
tesging

studyztp committed:
added simpoint support for set_se_workload and simulator

studyztp committed:
stdlib: take and create exit event for simpoint

Change-Id: I26b447801e22a51a182483faf878b580e53f5799

studyztp committed:
stdlib: kernal set workload takes simpoints

Change-Id: Id8e2c466571a63c7671b7eafab8080834cc4b44e

gzaets committed:
python: adding BBV generator config

Change-Id: I2b711fc5b065c6995b7a3dd9246f4b61e4711e86

studyztp committed:
stdlib: simulator() includes simpoint schedule functions

Change-Id: Ib5b99f5edff6b930400e2921f5ebb53934154197

studyztp committed:
stdlib: added simpoint class and simpoint exit event support

Change-Id: I1407882d4d8c120ea5ee2bad9f209f81bd50e8e3

studyztp committed:
Update kernel_disk_workload.py

studyztp committed:
Update src/cpu/base.cc

Co-authored-by: Jason Lowe-Power <jason@lowepower.com>

studyztp committed:
stdlib: support MAX_INSTS and schedule exit events

BaseCPU.py
Added scheduleOneMaxInsts

base.cc&base.hh
Moved original max instruction count event to scheduleOneMaxInsts
Removed support for SIMPOINT_END

se_workload.py
Added comments about the simpoint parameter
Removed some old simpoint comments

abstract_core.py&simple_core.py
Added set_one_max_insts function to schedule for MAX_INSTS
Added new parameters init for both scheduling functions to choose
when to schedule the events
Added pydoc comments
Removed init_simpoint function and end_point parameter

simpoint.py
Added license
Changed to using with statement to open files
Modified comments and variable names

exit_event.py
Added MAX_INSTS

simulation.py
Added schedule_max_insts function
Added pydoc comments for new scheduling functions

Change-Id: Ie655c95fd8bff3669c7ea06c14f69c0f758451aa

studyztp committed:
Update base.cc

studyztp committed:
stdlib: modified SimPoints supports

fs_checkpoint.py & fs_restore.py
two example scripts for using SimPoints related functions
00000000000000000000000000000000000000000000000000000000000000000000000000
BaseCPU.py, base.cc & base.hh
Changed the name of the schedule function for MAX_INSTS to
scheduleInstStopAnyThread

se_binary_workload.py
Added warning when using SimPoints functions with more than one core

abstract_core.py & simple_core.py
Added pydoc lines to explain the parameters for the SimPoints and
MAX_INSTS schedule functions
Changed the name for the MAX_INTS schedule function to
set_inst_stop_any_thread

simpoint.py
Added an empty line at the end of the file
Added usage note to ask for either paths or lists for both SimPoints and
weights

exit_event_generators.py
Removed unused libraries
Changed to using pathlib in the SimPoints checkpoint generator

simulator.py
Added warning when using SimPoints functions with more than one core

Removed the bbvprobe.py because it was a testing file

Change-Id: I1f353707cbeeb7f47d3df331a7e01df1807f7093

studyztp committed:
Update fs_restore.py

studyztp committed:
configs: added comments to the configs scripts

Change-Id: Iba78f2097eaecc96d9fb08041e8324442a4fddd8

studyztp committed:
Rename fs_checkpoint.py to simpoints-fs-checkpoint.py

studyztp committed:
Rename fs_restore.py to simpoints-fs-restore.py

studyztp committed:
stdlib, configs: changed examples from FS to SE

Fixed some grammar issues in comments and pydocs
Changed the example configs scripts to SE mode

Change-Id: I9e65a16c2415a2ae2d8dfad326010d1fe1bc14de

studyztp committed:
Rename simpoints-fs-checkpoint.py to simpoints-se-checkpoint.py

studyztp committed:
Rename simpoints-fs-restore.py to simpoints-se-restore.py

studyztp committed:
stdlib: fixed some style issues with simpoint.py

Change-Id: Ib25b369cf01384939db49a9a0b828e9f749dafe6

studyztp committed:
fixed a mistake line in the simpoint-se-restore.py

studyztp committed:
configs: removed commands

Change-Id: Id0a219c9fdf8c3c46f180aac2e8c4f40eb2182ec

studyztp committed:
stdlib, configs: added warmup support and comments in examples

simpoints-se-checkpoint.py & simpoints-se-restore.py
Changed cache hierarchy to NoCache() when taking checkpoints
Changed memory to simpler memory when taking checkpoints
Added warmup sample
Added comments for cache and memory

base.cc
Fixed style issue

simpoint.py
Added a new parameter to take in warmup interval
Added a function to setup warmup instructions for warmup support
Added a new object in the SimPoint class to store warmup list

Change-Id: I8f66d60b6ed92ac6d0e9841d854803c571a5d10c

studyztp committed:
rebase with the public gem5 develop

studyztp committed:
Delete testing.txt

studyztp committed:
Delete output.txt

studyztp committed:
congfigs: added a comment to explain the schedule_max_insts

studyztp committed:
stdlib: moved function from simple_core to base_cpu_core

BaseCPU.py
formatted with python black

abstract_core.py
formatted with python black

simple_core.py
removed simpoint related scheduling functions

base_cpu_core.py
added set_simpoint and set_inst_stop_any_thread

simpoint.py
changed a variable name 'inst_start' to 'start_inst'

This commit was tested with the simpoint config scripts in the
configs/example/gem5_library/checkpoints

Change-Id: I7e30ecc79db3f592e5f9f192c4b936edda0a698a

studyztp committed:
stdlib: formatted the abstract_core.py with python black

Change-Id: I2c1a57d43818f4e352be4143ef4cf8414953b17e

studyztp committed:
stdlib: moved simpoint.py from processor dir to utils

simpoints-se-checkpoint.py & simpoint-se-restore.py:
added explaination and usage on the top of the scripts
changed SimPoint module directory

SConscript:
changed SimPoint module directory

se_binary_workload.py:
changed SimPoint module directory

abstract_core.py:
fixed the format

simpoint.py:
moved dirctory from processor to utils

This commit is tested on the simpoint-se-checkpoint.py and
the simpoint-se-restore.py

Change-Id: Ib01f097660855a657bb1f504297ae37495813c7d

studyztp committed:
stdlib: fixed the error message of MAX_INSTS schedule function

Change-Id: Iffe00e8beea59f4bb35392c12d3173e8c5f71d6b

studyztp committed:
stdlib: added back some changes in simple_core.py

The rebase I've done a week ago didn't merge changes in the
simple_core.py.

studyztp committed:
stdlib: remove some default values and fix the fucntion names

simpoints-se-checkpoint.py:
Changed the generator name

se_binary_workload.py:
passing init value to the schedule_simpoint function

abstract_core.py:
added ":bool" to init

base_cpu_core.py:
removed default values for set_simpoint and set_inst_stop_any_thread

exit_event_generators.py:
change dump/reset and checkpoint generator from simpoint specific
to not simpoint specific

simulator.py:
removed default value for insts and start_insts for simpoint and
max-inst scheduling functions

simpoint.py:
changed the way to init warmup list
applied python black format

Change-Id: Ia83c2d6b68dce6ea034b53ab73754e4dfa2c48a8

BobbyRBruce commented on configs/example/gem5_library/checkpoints/simpoints-se-checkpoint.py:
```diff
@@ -0,0 +1,126 @@
+    simpoint_interval=1000000,
+    warmup_interval=1000000
+    # simpoint_file_path=Path("path/to/simpoints"),
+    # weight_file_path=Path("path/to/weights"),
```
I'd remove these before pushing to gerrit. They are not needed right now.

BobbyRBruce commented on configs/example/gem5_library/checkpoints/simpoints-se-restore.py:
```diff
@@ -0,0 +1,142 @@
+
+# schedule a MAX_INSTS exit event before the simulation begins
+# the schedule_max_insts function only schedule event when the instruction length is greater than 0
+simulator.schedule_max_insts(simpoint.get_warmup_list()[3], True)
```
What's special here about the `[3]`?

studyztp replied:
[3] is the index of the fourth SimPoint. I thought restoring a later SimPoint can be a better example because it would start from "print 12000" something, but it doesn't have to be [3]. It can be [0], [1], or [2] too. As long as the checkpoint file matches the index.

BobbyRBruce commented on configs/example/gem5_library/checkpoints/simpoints-se-restore.py:
```diff
@@ -0,0 +1,142 @@
+    simpoint_interval = 1000000,
+    warmup_interval = 1000000
+    # simpoint_file_path=Path("path/to/simpoints"),
+    # weight_file_path=Path("path/to/weights"),
```
I'd just remove these parameters for now.

BobbyRBruce commented on src/python/gem5/components/processors/abstract_core.py:
```diff
@@ -120,3 +120,31 @@ def get_mmu(self) -> BaseMMU:
         raise NotImplementedError
+
+    @abstractmethod
+    def set_simpoint(self, inst_starts, init: bool) -> None:
```
`inst_starts` is not typed. IShould be `inst_starts: List[int]`

BobbyRBruce commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -185,6 +188,38 @@ def unique_exit_event():
         self._checkpoint_path = checkpoint_path
 
+    def schedule_simpoint(
+        self, simpoint_start_insts, schedule_at_init: bool = False
```
`simpoint_start_insts` needs typed.

BobbyRBruce reviewed (commented):
This is very good work. I've got very little problem with it outside a few minor comments and one question.

powerjg commented on src/python/SConscript:
```diff
@@ -211,6 +211,8 @@ PySource('gem5.components.processors',
     'gem5/components/processors/simple_switchable_processor.py')
 PySource('gem5.components.processors',
     'gem5/components/processors/switchable_processor.py')
+PySource('gem5.utils',
+    'gem5/utils/simpoint.py')
```
```suggestion
PySource('gem5.utils', 'gem5/utils/simpoint.py')
```

powerjg commented:
There's a couple small comments from Bobby. Otherwise, we just need to partition this into the commits you suggested, Zhantong, and then submit to gerrit!

studyztp committed:
stdlib: added type def and fixed some format

simpoints-se-checkpoint.py&
simpoints-se-restore.py:
removed comment about simpoint file path

SConscript:
fixed the format

abstract_core.py &
base_cpu_core.py &
simulator.py:
added type for start_insts and insts for scheduling functions

simulator.py:
applied the patch for simQuantum
